lib/web_ui/test/engine/platform_dispatcher_test.dart

```diff
@@ -111,5 +111,52 @@ void testMain() {
       root.style.fontSize = null;
       expect(findBrowserTextScaleFactor(), 1.0);
     });
+
+    test(
+        'calls onTextScaleFactorChanged when the <html> element\'s font-size changes',
+        () async {
+      final html.Element root = html.document.documentElement!;
+      final String oldFontSize = root.style.fontSize;
+
+      addTearDown(() {
+        root.style.fontSize = oldFontSize;
+      });
+
+      root.style.fontSize = '16px';
+
+      bool callsCallback = false;
+      ui.PlatformDispatcher.instance.onTextScaleFactorChanged = () {
+        callsCallback = true;
+      };
+
+      Future<void> waitUntilCalled() {
+        final Completer<void> completer = Completer<void>();
+
+        void check() {
+          if (callsCallback == true) {
+            completer.complete();
+          } else {
+            Timer(Duration.zero, check);
+          }
+        }
+
+        check();
+        return completer.future;
+      }
+
+      root.style.fontSize = '20px';
+      await waitUntilCalled();
```

Comment:
This part of the test seems problematic. To allow it to pass, I must wait until the `MutationObserver` has had time to react to the change before I call `expect`, meaning that if I break the observer then the test will hang here for 30 seconds until it automatically times out. Obviously, it would be better to have an immediate fail. What's the best way to let the test fail without violating the [style guide](https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#never-check-if-a-port-is-available-before-using-it-never-add-timeouts-and-other-race-conditions)?

Reply:
With the way the test is structured, the `expect(callsCallback, true);` is not doing anything because we can't reach it unless `callsCallback` is already true.

Do you know how quickly the browser fires the mutation observer? If it happens in the next microtask, then maybe you can just have a `Timer(Duration.zero)` and then immediately check for `callsCallback`?

Reply:
The `expect` call is redundant and the primary goal is to make it relevant because right now it runs prematurely without `waitUntilCalled` or a similar delay.

I tested several alternatives. For instance, wrapping `expect(isCalled, true)` in a `Future()` ostensibly makes the test work without requiring a delay function. I ran this edition dozens of times locally against various scenarios and it always gave the right output. However, I'm uncertain about this solution. Does it's efficacy make sense, or am I just getting lucky?

Reply:
What `MutationObserver` documentation I could find suggests that the browser does call observer's callback by queuing a microtask. So, as long as no others are scheduled in between, everything should "happen in the next microtask."

Furthermore, it appears that `Future()` uses `Timer.run()` under the hood, which then returns `Timer(Duration.zero, callback)`, meaning the solution I tried should be equivalent to the one you suggested. Should I go ahead and commit?

Reply:
With the latest changes you made to the test, it looks good to me!